rstudio/rstudio
# Fix recent items breaking due to menu separators fix

### Intent
Undo the damage from fixing the menu separators. This is a refactor of how the menus are constructed and the menu model. I took a quick peek at the menus and everything looked the same (as far as I can tell from my limited use of the IDE). Added a few unit tests to cover the scenarios where the menus are updated in different ways: changing a menu item label/visibility, full menu rebuild due to reloading the UI, and updating shortcuts.

I've also updated the lint warnings about unused variables. There's an ignore added so that variables prefixed with `_` are ignored for this rule.

### Approach
Change the menu model as `MenuItemConstructorOptions`. This makes it easy to pass a template to `Menu.buildFromTemplate()` to create the application menu. The menu model is copied into a new template with unnecessary menu items and separators. The model will retain the state of menu items like visibility and the menu item order. GWT only needs to make calls to Electron to update commands and eventually call `menu_commit_command_shortcuts` to refresh the menu.

### Automated Tests
There are a few more unit tests to cover changing visibility, labels, shortcuts, and completely rebuilding the menu. 

### QA Notes
* Changing projects should update the recent projects submenu
* Opening files using the Files pane should update the recent files submenu
* Changing the project build options should rebuild the entire menu and show the new options
* Reloading the UI using the Help > Diagnostics > Reload UI action should rebuild the menu

### Checklist
